### PR TITLE
Add functor examples

### PR DESCRIPTION
**Pull request by vvp**
This PR adds computation abstract examples for functors.

### CODE REVIEW

File: functors/README.md [hunk 1]
```diff
@@ -0,0 +1,13 @@
+# Functors
+
+This directory contains examples of _functors_ with various strategies for distributed evaluation.
+
+[Locally evaluated identity functor](identity-functor.amb), roughly equivalent to following JS code:
+
+```javascript
+let identity = (value) => () => ({ value })
+let map_identity = () => (id, func) => identity(func(id().value))
```
Comment (haadcode) [inline functors/README.md]: Can we rename "func" to something else to not confuse the argument name with func primitive in the encoding? Eg. "f" of "fn"?

**Review by haadcode — COMMENTED** [review]
The end of the encoding trips me as there's no encoding for "identity". Can we make that explicit and add it?

File: functors/identity-functor.amb [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+  call[
+    out program.in map_identity.open_.return[open_.in program.in func]
+  ]|
+  call[
+    out program.in string_length.open_.return[open_.in program.in func.in func]
+  ]|
+  func[in_ map_identity.open map_identity.(
+    arg[
+      string[hello[]]|
+      in id.open_]|
```
Comment (haadcode) [inline functors/identity-functor.amb]: Could we wrap the arg to an `identity` and `open` it once in `id` to address the earlier feedback on lacking the explicit `identity` here?

**Comment by haadcode** [comment]
*removed as obsolete*

File: functors/README.md [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+
+This directory contains examples of _functors_ with various strategies for distributed evaluation.
+
+[Locally evaluated identity functor](identity-functor.amb), roughly equivalent to following JS code:
+
+```javascript
+let identity = (value) => () => ({ value })
+let map_identity = () => (id, fn) => identity(fn(id().value))
+let string_length = (str) => str.length
+let program = map_identity()(identity("hello"), string_length)
```
Comment (haadcode) [inline functors/README.md]: I believe this should be `let program = () => map_identity(...`

**Comment by haadcode** [comment]
> Here's another version that imo clears up some confusion

The general reasoning/idea here is that functions with one arg don't get named params, functions with *n* arguments get passed (and handle internally the routing of) named params.

**Comment by haadcode** [comment]
> The general reasoning/idea here is that functions with one arg don't get named params, functions with n arguments get passed (and handle internally the routing of) named params.

If we want to pass `arg`s from `program` to the `map_identity`, we could add another set of wrappers for `id` and `fn` in `map_identity` (that then pass them into the `arg`s inside `map_identity`)

**Comment by haadcode** [comment]
> If we want to pass args from program to the map_identity, we could add another set of wrappers for id and fn in map_identity (that then pass them into the args inside map_identity)

Updated the encoding above to do ^.

**Comment by haadcode** [comment]
After discussion on Slack, ignore the above comments. The encoding provided by the original PR is 👌

File: functors/README.md [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+    ]|
+    open return.open_
+  )
+]|
+map_identity[
+  in_ call.open call.(
+    func[
+      id[in_ arg.open arg.open identity.in arg.open_]|
+      arg[in_ id.open id.in func.in arg.open_]|
+      func[in_ func.in_ arg.open func.in identity.open_.open func]|
```
Comment (haadcode) [inline functors/README.md:40]: Just to understand this: why `open func` after `open_`? Iirc we discussed that `open_` is always the last action.

**Comment by haadcode** [comment]
LGTM 👍 Had one question, see above.

File: functors/README.md (hunk 4 above)
Comment (vvp) [reply functors/README.md:40]: Good catch, I'll change that. It's nice though that with these hierarchical evaluation contexts, both versions do work ie. the converge the same even though at this particular level of evaluation the ordering is different.

Comment (haadcode) [reply functors/README.md:40]: Yup, realized that too. And we actually discussed it (and iirc define in the paper) that "the remaining caps/ambients will be inhereted by the opening ambient" or smth like that. So this is a great example of it :)

**Pull request merged**